### PR TITLE
Check if driver support DPAS and 2D block operations

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Module attributes `triton_gpu.support_sg_2d_block` and `triton_gpu.support_dpas` are added to the module if the driver supports the corresponding OpenCL extension.
This way is preferred instead of checking if a driver is LTS, since the OpenCL extension can also be unsupported for non-LTS driver, and checking driver version is error prone.

### CODE REVIEW

File: test/Conversion/intel/arith_to_llvm.mlir [hunk 1]
```diff
@@ -12,7 +12,7 @@
 // CHECK-LABEL:   llvm.func spir_kernelcc @float_to_bfloat_conversion(
 // CHECK-SCALAR:                                             %[[VAL_0:.*]]: !llvm.struct<(f32, f32, f32, f32)>) -> !llvm.struct<(bf16, bf16, bf16, bf16)>
 // CHECK-VECTOR:                                             %[[VAL_0:.*]]: vector<32xf32>) -> vector<32xbf16>
-module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 4 : i32} {
+module attributes {"triton_gpu.support_sg_2d_block" = 1 : i1, "triton_gpu.support_dpas" = 1 : i1, "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 4 : i32} {
```
Comment (victor-eds) [inline test/Conversion/intel/arith_to_llvm.mlir:15]:
```suggestion
module attributes {"triton_gpu.support_sg_2d_block", "triton_gpu.support_dpas", "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 4 : i32} {
```

I think we can make this a `UnitAttr` so we just need to check whether it's present

File: third_party/intel/triton_xpu.cc [hunk 2]
```diff
@@ -94,12 +94,18 @@ void init_triton_intel(py::module &&m) {
...
     auto i1_ty = mlir::IntegerType::get(mod->getContext(), 1);
+    // FIXME: Use SYCL runtime to query supported OpenCL extensions, instead of
+    // checking driver version.
     if (isLTS)
       mod->setAttr("triton_gpu.is_lts", mlir::IntegerAttr::get(i1_ty, 1));
+    if (supportSG2DBlock)
+      mod->setAttr("triton_gpu.support_sg_2d_block",
+                   mlir::IntegerAttr::get(i1_ty, 1));
+    if (supportDPAS)
+      mod->setAttr("triton_gpu.support_dpas", mlir::IntegerAttr::get(i1_ty, 1));
```
Comment (victor-eds) [inline third_party/intel/triton_xpu.cc:108]:
```suggestion
    Builder b(mod);
    if (supportSG2DBlock)
      mod->setAttr("triton_gpu.support_sg_2d_block", b.getUnitAttr());
    if (supportDPAS)
      mod->setAttr("triton_gpu.support_dpas", b.getUnitAttr());
```

**Review by victor-eds — COMMENTED** [review]
Nice improvement! Should the prefix be `triton_intel_gpu` instead of `triton_gpu`? Should we have the attribute name be queried from `TritonIntelGPUDialect` instead of being constants in the code?

File: third_party/intel/backend/compiler.py [hunk 3]
```diff
@@ -153,16 +153,16 @@ def make_ttir(mod, metadata, opt):
         return mod
 
     @staticmethod
-    def make_ttgir(mod, metadata, opt, device_arch):
+    def make_ttgir(mod, metadata, opt, device_arch, support_sg_2d_block, support_dpas):
         is_lts = Version(metadata["target"].arch["driver_version"]) == Version("1.3.27642")
-        if (not is_lts and os.getenv("TRITON_INTEL_ENABLE_BLOCK_PTR", "0") == "1"):
+        intel.set_device_properties(mod, is_lts, support_sg_2d_block, support_dpas)
```
Comment (LiyangLingIntel) [inline third_party/intel/backend/compiler.py:158]: As we can query if features are supported ornot, is `is_lts` still needed for the passes? Are there any features not covered?

File: test/Conversion/intel/arith_to_llvm.mlir (hunk 1 above)
Comment (whitneywhtsang) [reply test/Conversion/intel/arith_to_llvm.mlir]: done

File: third_party/intel/triton_xpu.cc (hunk 2 above)
Comment (whitneywhtsang) [reply third_party/intel/triton_xpu.cc]: done

**Comment by whitneywhtsang** [comment]
> Nice improvement! Should the prefix be triton_intel_gpu instead of triton_gpu? Should we have the attribute name be queried from TritonIntelGPUDialect instead of being constants in the code?

Done.

File: third_party/intel/backend/compiler.py [hunk 4]
```diff
@@ -153,16 +153,16 @@ def make_ttir(mod, metadata, opt):
         return mod
 
     @staticmethod
-    def make_ttgir(mod, metadata, opt, device_arch):
+    def make_ttgir(mod, metadata, opt, device_arch, support_sg_2d_block, support_dpas):
```
Comment (LiyangLingIntel) [inline third_party/intel/backend/compiler.py:156]: How about use `self.properties` here and get the actual properties in `intel.set_device_properties`? In this way, when there are further property changes, we do not need to change the interface back and forth.

Comment (whitneywhtsang) [reply third_party/intel/backend/compiler.py]: done

**Review by LiyangLingIntel — APPROVED** [review]
LGTM

**Review by victor-eds — COMMENTED** [review]
LGTM after small change

File: third_party/intel/include/TritonIntelGPUToLLVM/TypeConverter.h [hunk 5]
```diff
@@ -18,7 +18,7 @@ class TritonIntelGPUToLLVMTypeConverter : public TritonGPUToLLVMTypeConverter {
   using TypeConverter::convertType;
 
   TritonIntelGPUToLLVMTypeConverter(
-      MLIRContext *ctx, LowerToLLVMOptions &option, bool isLTSDriver,
+      MLIRContext *ctx, LowerToLLVMOptions &option, bool isBlockPtrPathEnabled,
```
Comment (etiotto) [inline third_party/intel/include/TritonIntelGPUToLLVM/TypeConverter.h:21]: Lets call it experimental path (Block pointers are a language features)

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp [hunk 6]
```diff
@@ -79,9 +79,14 @@ struct ConvertTritonGPUToLLVM
     intel::TritonGPUToLLVMPipelineManager pipelineManager(mod, context);
     mlir::LowerToLLVMOptions option(context);
     option.overrideIndexBitwidth(32);
-    bool isLTSDriver = mod->hasAttr("triton_gpu.is_lts");
+    bool isBlockPtrPathEnabled =
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:82]: Lets call `isExperimentalPath` (everwhere)

File: third_party/intel/include/TritonIntelGPUToLLVM/TypeConverter.h (hunk 5 above)
Comment (whitneywhtsang) [reply third_party/intel/include/TritonIntelGPUToLLVM/TypeConverter.h]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp (hunk 6 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp]: done

**Review by chengjunlu — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Pull request merged**